prefect_kubernetes/utilities.py

```diff
@@ -180,3 +185,126 @@ def _slugify_label_value(value: str, max_length: int = 63) -> str:
+            try:
+                for event in self.watch.stream(func, *args, **kwargs):
+                    # check that we want to and can track this object
+                    if (
+                        cache is not None
+                        and isinstance(event, dict)
+                        and "object" in event
+                    ):
+                        uid = event["object"].metadata.uid
+                        if uid not in cache:
+                            cache[uid] = None
+                            yield event
+                    else:
+                        yield event
+                else:
+                    # Case: we've finished iterating
+                    keep_streaming = False
+            except self.reconnect_exceptions:
+                # Case: We've hit an exception we're willing to retry on
+                if self.logger:
+                    self.logger.error("Unable to connect, retrying...", exc_info=True)
+                time.sleep(1)
+            except Exception:
+                # Case: We hit an exception we're unwilling to retry on
+                if self.logger:
+                    self.logger.exception(
+                        f"Unexpected error while streaming {func.__name__}"
+                    )
+                keep_streaming = False
+                self.stop()
+                raise
+
+        self.stop()
+
+    def api_object_stream(self, func: Callable, *args, **kwargs):
+        """
+        Create a FIFOCache to maintain a record of API objects that have been
+        seen. This is useful because `_stream` will reconnect a stream on
+        `self.reconnect_exceptions` and on reconnect it will restart streaming all
+        objects. This cache prevents the same object from being yielded twice.
+
+        Args:
+            func: A Kubernetes client function to call which produces a stream of API o
+                bjects
+            *args: Positional arguments to pass to `func`
+            **kwargs: Keyword arguments to pass to `func`
+
+        Returns:
+            An iterator of API objects
+        """
+        cache = FIFOCache(maxsize=self.max_cache_size)
+        yield from self._stream(cache, func, *args, **kwargs)
+
+    def log_stream(self, func: Callable, *args, **kwargs):
+        """
+        Stream logs genererated by a function in a way that will reconnect the
+        stream on `self.reconnect_exceptions`. On reconnect, the stream will
+        restart streaming starting from the beginning of the log's history. It
+        is possible for duplicate logs to be yielded.
+
+        Args:
+            func: A Kubernetes client function to call which produces a stream
+                of logs
+            *args: Positional arguments to pass to `func`
+            **kwargs: Keyword arguments to pass to `func`
+
+        Returns:
+            An iterator of log
+        """
+        yield from self._stream(None, func, *args, **kwargs)
```

Comment:
For my curiosity, why can we not reliably deduplicate logs?

Reply:
The client log functions return plain strings, and plain strings by themselves are ambiguous - if a pod has repetitive output it's impossible to tell whether it's new log with the same output or a log that we've seen before. 

If someone does `echo hi` 5 times, deduplication would return only a single log entry, which seemed more misleading than the occasion duplicated log stream.

Reply:
Gotcha, that makes sense!